From f7d3a0dfa7aed8472d68af2df1e1e742e6c612d3 Mon Sep 17 00:00:00 2001
From: Tom Hughes <tom@compton.nu>
Date: Tue, 17 Nov 2015 17:36:16 +0000
Subject: [PATCH] Remove size limit when expanding macros

This removes a seemingly undocumented, and not even well defined, limit
on the size of a macro expansion when parsing a spec file.

[lkardos@redhat.com: created new funtion expandMacrosU() (Unlimited)
instead of modifying expandMacros() in order not to change API/ABI]

Signed-off-by: Lubos Kardos <lkardos@redhat.com>
(cherry picked from commit 61838b0fdacb71a881baac164043b8e40ddfbec5)

[Quentin: conflicts due to a few upstream commits missing that are
shuffling around various helper functions, non-exhaustive list:
 - b823075695ec ("Separate internal vs external macro add and delete
 functions")
 - acead93927e4 ("Group exported functions to end of macro.c, no functional
 changes")
 - c9e34b257ea5 ("Drop slen from builtin macro parser function")
 - 1036601e3eba ("Add and use dummy macro context lock acquire + release
 functions")
 - 09499d994b9c ("Add and use internal (lockless) function for copying
 macro contexts")
 - 71e2d38e3d75 ("Make macros thread-safe ... sort of")
 - 784dfb421924 ("Port %{load:<path>} macro from rpm5.org")
 - d4ab1d82d788 ("Macro-expand %{load:...} argument to make the thing more
 useful...")
 - 54f24ec5486b ("Fix reading a memory right after the end of an allocated
 area.")
 - e7a608809317 ("Warn on empty Warn-on-empty-files-f-manifest-filesiles -f
 manifest files")
 - 2a3f49585e5b ("Warn when macro is expanded in comment.")
]

Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>
---
 build/files.c    |  4 +++-
 build/pack.c     |  6 ++++--
 rpmio/macro.c    | 15 +++++++++++++++
 rpmio/rpmmacro.h | 12 ++++++++++++
 4 files changed, 34 insertions(+), 3 deletions(-)

diff --git a/build/files.c b/build/files.c
index d8fe5384aa67..ecca782ee01c 100644
--- a/build/files.c
+++ b/build/files.c
@@ -1664,11 +1664,13 @@ static rpmRC readFilesManifest(rpmSpec spec, Package pkg, const char *path)
 
     while (fgets(buf, sizeof(buf), fd)) {
 	handleComments(buf);
-	if (expandMacros(spec, spec->macros, buf, sizeof(buf))) {
+	char *expanded = expandMacrosU(spec, spec->macros, buf);
+	if (expanded == NULL) {
 	    rpmlog(RPMLOG_ERR, _("line: %s\n"), buf);
 	    goto exit;
 	}
 	argvAdd(&(pkg->fileList), buf);
+	free(expanded);
     }
 
     if (ferror(fd))
diff --git a/build/pack.c b/build/pack.c
index da5eae928adb..93ce4ac9c034 100644
--- a/build/pack.c
+++ b/build/pack.c
@@ -88,11 +88,13 @@ static rpmRC addFileToTag(rpmSpec spec, const char * file,
     }
 
     while (fgets(buf, sizeof(buf), f)) {
-	if (expandMacros(spec, spec->macros, buf, sizeof(buf))) {
+	char *expanded = expandMacrosU(spec, spec->macros, buf);
+	if (expanded == NULL) {
 	    rpmlog(RPMLOG_ERR, _("%s: line: %s\n"), fn, buf);
 	    goto exit;
 	}
-	appendStringBuf(sb, buf);
+	appendStringBuf(sb, expanded);
+	free(expanded);
     }
     headerPutString(h, tag, getStringBuf(sb));
     rc = RPMRC_OK;
diff --git a/rpmio/macro.c b/rpmio/macro.c
index dc4de4a47920..74840f1a8118 100644
--- a/rpmio/macro.c
+++ b/rpmio/macro.c
@@ -1267,6 +1267,21 @@ int expandMacros(void * spec, rpmMacroContext mc, char * sbuf, size_t slen)
     return rc;
 }
 
+char *expandMacrosU(void * spec, rpmMacroContext mc, char * sbuf)
+{
+    char *target = NULL;
+    int rc;
+
+    rc = doExpandMacros(mc, sbuf, &target);
+
+    if (rc) {
+	free(target);
+	target = NULL;
+    }
+
+    return target;
+}
+
 void
 addMacro(rpmMacroContext mc,
 	const char * n, const char * o, const char * b, int level)
diff --git a/rpmio/rpmmacro.h b/rpmio/rpmmacro.h
index 765c78c6dd6f..a4884b0ae04d 100644
--- a/rpmio/rpmmacro.h
+++ b/rpmio/rpmmacro.h
@@ -65,6 +65,18 @@ int	expandMacros	(void * spec, rpmMacroContext mc,
 				char * sbuf,
 				size_t slen);
 
+/** \ingroup rpmmacro
+ * Expand macro into buffer.
+ * @deprecated Use rpmExpand().
+ * @todo Eliminate from API.
+ * @param spec		cookie (unused)
+ * @param mc		macro context (NULL uses global context).
+ * @retval sbuf		input macro to expand
+ * @return		macro expansion (malloc'ed) or NULL on failure
+ */
+char	*expandMacrosU	(void * spec, rpmMacroContext mc,
+				char * sbuf);
+
 /** \ingroup rpmmacro
  * Add macro to context.
  * @deprecated Use rpmDefineMacro().
